Update codebase to use logging instead of print

ankona opened the pull request:
This change replaces all usage of `print()` with python `logging.Logger` and updates tests as-needed.

- [X] Closes #88 
- [X] Tests added
- [X] Tests passing
- [X] Full type hint coverage
- [ ] Changes are documented in `docs/releases.rst`
- [ ] New functions/methods are listed in `api.rst`
- [ ] New functionality has documentation

ankona committed:
Update codebase to use logging instead of print

ankona committed:
Adjust typehint to allow partial-second update durations

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1039,29 +1039,29 @@ def setup(
         infostr = f"Configuring {self.__class__.__name__}"
-        print(infostr + "\n" + "-" * len(infostr))
-        print(f"Setting up {self.codebase.__class__.__name__}...")
+        (infostr + "\n" + "-" * len(infostr))
```
Fixed improper removal of print stmt

ankona committed:
Formatting

ankona committed:
Fix typo removing log/print stmt

ankona commented on cstar/system/scheduler.py:
```diff
@@ -447,7 +449,7 @@ def global_max_cpus_per_node(self) -> Optional[int]:
         )
         if result.returncode != 0:
-            print(f"Error querying node property. STDERR: {result.stderr}")
+            self.log.info(f"Error querying node property. STDERR: {result.stderr}")
```
Replaced `.info` w/ `.error` in messages indicating "error" in the message

ankona commented on cstar/tests/unit_tests/base/test_external_codebase.py:
```diff
@@ -28,20 +33,20 @@ def _codebase_adjustments(self):
 
 @pytest.fixture
-def generic_codebase():
+def generic_codebase(log: logging.Logger):
```
Fixture `def log() -> logging.Logger:` already added in `/cstar/tests/unit_tests/conftest.py`

ankona committed:
remove commented-out code

ankona committed:
avoid logging full stack trace on user interrupted process

ScottEilerman reviewed (approved):
I didn't dive into all the diffs, but the overall approach in log.py, and usage as a class mixin or module-level global looks good to me, and is consistent with other projects I've worked on. I don't have a good feel yet for what we want as the default format (I tend to like function name and line number, and levelname right after the time, but that's just personal preference, and maybe is less appropriate for an open-source library compared to the stuff I've worked on in the past). Anyway, we can always update it later too.

ScottEilerman commented on cstar/base/log.py:
```diff
@@ -0,0 +1,41 @@
+        return logger
+
+    fmt = format or DEFAULT_LOG_FORMAT
+    logging.basicConfig(level=logging.DEBUG, format=fmt)
```
```suggestion
    logging.basicConfig(level=level, format=fmt)
```
Looks like the passed in level never gets used? And I guess that is where we can set the default -- do we want debug or info?

ankona replied:
derp. edit fail. fixed.

ScottEilerman reviewed (changes requested):
I didn't dive into all the diffs, but the overall approach in log.py, and usage as a class mixin or module-level global looks good to me, and is consistent with other projects I've worked on. I don't have a good feel yet for what we want as the default format (I tend to like function name and line number, and levelname right after the time, but that's just personal preference, and maybe is less appropriate for an open-source library compared to the stuff I've worked on in the past). Anyway, we can always update it later too.

ankona committed:
fix bug ignoring user-supplied log level parameter

ankona committed:
docstrings, log fmt, improve caplog filtering

dafyddstephenson commented:
This is now blocked by #240 . I'd like to see the impact of the logging on the example notebooks before merging, but they won't run properly again until #240 has been addressed.

dafyddstephenson commented on cstar/tests/unit_tests/base/test_additional_code.py:
```diff
@@ -774,7 +775,7 @@ def test_get_with_template_files(self, local_additional_code):
         local_additional_code.get("/mock/local/dir")
 
-        print(f"DEBUG: All copy calls: {self.mock_copy.call_args_list}")
+        log.debug(f"DEBUG: All copy calls: {self.mock_copy.call_args_list}")
```
oh man this should not have been here 😅 my bad. Whatever I was debugging is presumably now resolved so this can go

ankona replied:
removed

dafyddstephenson commented on cstar/tests/unit_tests/base/test_input_dataset.py:
```diff
@@ -818,7 +827,7 @@ def test_local_hash_single_file(self, local_input_dataset):
 
         # Debug: Confirm resolve was invoked
-        print(f"Resolved path: {local_input_dataset.working_path.resolve()}")
+        log.info(f"Resolved path: {local_input_dataset.working_path.resolve()}")
```
Again this seems to be sloppy past clean-up by me and can go

ankona replied:
removed

dafyddstephenson commented on cstar/tests/unit_tests/roms/test_roms_simulation.py:
```diff
@@ -855,8 +859,8 @@ def test_tree(self, example_roms_simulation):
-        # print(sim.tree())
-        # print(expected_tree)
+        log.debug(sim.tree())
+        log.debug(expected_tree)
```
safe to delete, more tidying failures by me

ankona replied:
removed

dafyddstephenson commented on cstar/base/external_codebase.py:
```diff
@@ -245,7 +246,7 @@ def handle_config_status(self) -> None:
             case 2:
                 head_hash = _get_repo_head_hash(local_root)
-                print(
+                self.log.info(
```
I think the correct approach here is to roll this whole block into the `input` call below, rather than making it a log entry

ankona replied:
Reverting these changes - it's meant to be user-facing output, not logs.

dafyddstephenson commented on cstar/base/external_codebase.py:
```diff
@@ -266,13 +267,13 @@ def handle_config_status(self) -> None:
                     / f"externals/{self.repo_basename}"
                 )
-                print(
+                self.log.warning(
```
Again, this is part of the `input` prompt and should ideally be combined with that

ankona replied:
reverted!

ScottEilerman commented:
> This is now blocked by #240 . I'd like to see the impact of the logging on the example notebooks before merging, but they won't run properly again until #240 has been addressed.

@dafyddstephenson Can't we run everything in the notebook up to the actual ROMS execution step? Or even run that and let it fail and see how it looks in the new format.

dafyddstephenson commented:
> > This is now blocked by #240 . I'd like to see the impact of the logging on the example notebooks before merging, but they won't run properly again until #240 has been addressed.
> 
> @dafyddstephenson Can't we run everything in the notebook up to the actual ROMS execution step? Or even run that and let it fail and see how it looks in the new format.

Lol apparently I can edit your comments, meant to click reply:

We can, but the notebooks will need to be updated and working in line with the changes in this PR before it can be merged, which is not currently possible. Nora is working on the PR on the `roms-tools` side, this should be back to normal in a few hours. 

Also a general comment that issues of this nature can be avoided by fixing the ROMS checkout target (currently just `main`) in the example, but while we're in a state of fluid development with no users this just makes things easier.

dafyddstephenson commented on cstar/base/input_dataset.py:
```diff
@@ -249,7 +250,9 @@ def get(self, local_dir: str | Path) -> None:
 
         if (self.exists_locally) and (self.working_path == target_path):
-            print(f"Input dataset already exists at {self.working_path}, skipping.")
+            self.log.warning(
```
I think your previous implementation with `info` was more appropriate here.

ankona replied:
swapped to info

dafyddstephenson commented on cstar/base/log.py:
```diff
@@ -0,0 +1,41 @@
+        return logger
+
+    fmt = format or DEFAULT_LOG_FORMAT
+    logging.basicConfig(level=level, format=fmt)
```
I'm new to python logging and need some reassurance on this one - is it safe to reconfigure the root logger with every call? or perhaps this is a temporary measure until we assign handlers?

ankona replied:
The root logger isn't reconfigured when calling `basicConfig` again. I had the same thought and couldn't remember so I looked it up [here](https://docs.python.org/3/library/logging.html#logging.basicConfig).

```
This function does nothing if the root logger already has handlers configured, unless the keyword argument force is set to True.
```

dafyddstephenson commented on cstar/execution/handler.py:
```diff
@@ -126,13 +128,13 @@ def updates(self, seconds: int = 10, confirm_indefinite: bool = True):
                 self.status == ExecutionStatus.CANCELLED and self.output_file.exists()
             ):
-                print(f"See {self.output_file.resolve()} for job output")
+                self.log.warning(f"See {self.output_file.resolve()} for job output")
```
I think perhaps this should be restructured such that the warning message is first constructed in two steps (with the second string conditionally appended) and then issued as a single warning

ankona replied:
Good idea. Modified to be a single output

dafyddstephenson replied:
looks good - I'll be a pedant and suggest `warning_msg` over `error_msg` but I won't lose sleep over it

dafyddstephenson commented on cstar/execution/handler.py:
```diff
@@ -157,8 +159,8 @@ def updates(self, seconds: int = 10, confirm_indefinite: bool = True):
                         time.sleep(0.1)  # 100ms delay between updates
         except KeyboardInterrupt:
-            print("\nLive status updates stopped by user.")
+            self.log.warning("Live status updates stopped by user.")
```
Again I think `info` is more appropriate here

ankona replied:
swapped to info

dafyddstephenson commented on cstar/execution/local_process.py:
```diff
@@ -226,7 +226,7 @@ def cancel(self):
                 self._drop_process()
         else:
-            print(f"Cannot cancel job with status '{self.status}'")
+            self.log.warning(f"Cannot cancel job with status '{self.status}'")
```
suggest `info`

ankona replied:
swapped to info

dafyddstephenson commented on cstar/execution/local_process.py:
```diff
@@ -240,4 +240,6 @@ def wait(self):
             self._process.wait()
         else:
-            print(f"cannot wait for process with execution status '{self.status}'")
+            self.log.warning(
```
suggest `info`

ankona replied:
swapped to info

dafyddstephenson commented on cstar/execution/scheduler_job.py:
```diff
@@ -726,7 +728,7 @@ def cancel(self):
 
         if self.status not in {ExecutionStatus.RUNNING, ExecutionStatus.PENDING}:
-            print(f"Cannot cancel job with status '{self.status}'")
+            self.log.warning(f"Cannot cancel job with status '{self.status}'")
```
suggest `info`

ankona replied:
swapped to info

dafyddstephenson commented on cstar/execution/scheduler_job.py:
```diff
@@ -963,7 +965,7 @@ def cancel(self):
             ExecutionStatus.HELD,
         }:
-            print(f"Cannot cancel job with status {self.status}")
+            self.log.warning(f"Cannot cancel job with status {self.status}")
```
suggest `info`

ankona replied:
swapped to info

dafyddstephenson commented on cstar/roms/input_dataset.py:
```diff
@@ -157,7 +160,9 @@ def get(
 
         if (self.exists_locally) and (working_path_parent == local_dir):
-            print(f"Input dataset already exists in {working_path_parent}, skipping.")
+            self.log.warning(
```
suggest `info`

ankona replied:
swapped to info

dafyddstephenson commented on cstar/roms/simulation.py:
```diff
@@ -924,9 +923,8 @@ def partitioned_files_to_runtime_code_string(input_dataset):
-        runtime_code_modifications[nl_idx]["__FORCING_FILES_PLACEHOLDER__"] = (
-            runtime_code_forcing_str.lstrip()
-        )
+        namelist = runtime_code_modifications[nl_idx]
```
suggest renaming this local var. I realise there are a lot of `nl`s here, but I've been trying to move away from calling the file in question here a `namelist` as the file we're modifying isn't technically a namelist (even though it's used in a similar way): its formatting is unique to ROMS.

ankona replied:
renamed to `formatting`. example:

```py
# Time step entry
formatting = runtime_code_modifications[nl_idx]
formatting["__TIMESTEP_PLACEHOLDER__"] = self.discretization.time_step
```

dafyddstephenson commented on cstar/system/scheduler.py:
```diff
@@ -447,7 +449,7 @@ def global_max_cpus_per_node(self) -> Optional[int]:
         )
         if result.returncode != 0:
-            print(f"Error querying node property. STDERR: {result.stderr}")
+            self.log.error(f"Error querying node property. STDERR: {result.stderr}")
```
Could you merge #235 ? Not sure how this will look after

ankona replied:
Reverted logging calls for these properties where the output is intended to be raw CLI output (not logs)

dafyddstephenson commented on cstar/tests/unit_tests/base/test_discretization.py:
```diff
@@ -8,10 +9,10 @@ def discretization():
     """Test the attributes were set correctly."""
     assert discretization.time_step == 3
-    print(discretization)
+    log.info(discretization)
```
I think this is another relic of past debugging. Sorry these seem to be all over the place in the tests.

ankona replied:
removed

dafyddstephenson commented on cstar/tests/unit_tests/execution/test_slurm_job.py:
```diff
@@ -407,7 +408,7 @@ def test_cancel(self, mock_status, mock_subprocess, tmp_path):
 
         # Log mock call arguments
-        print(f"Mock call args after failure: {mock_subprocess.call_args_list}")
+        log.info(f"Mock call args after failure: {mock_subprocess.call_args_list}")
```
likely more leftovers, can't honestly remember

ankona replied:
removed

dafyddstephenson commented on cstar/tests/unit_tests/roms/test_roms_external_codebase.py:
```diff
@@ -140,7 +141,7 @@ def test_get_success(self, roms_codebase, tmp_path):
 
         ## Check that subprocess.run was (mock) called twice for `make nhmg` and `make Tools-Roms`
-        print(self.mock_subprocess_run.call_args_list)
+        log.info(self.mock_subprocess_run.call_args_list)
```
Again I think this is likely leftover debugging stuff (?)

ankona replied:
removed

dafyddstephenson reviewed (changes requested):
Monster effort for a second PR, thanks a tonne for getting this done. Some suggestions, primarily around levels for different messages. Safe to assume print statements in unit tests that aren't part of assertions are leftovers from past debugging, can either be kept as debug or deleted entirely - apologies for my sloppiness there!

ankona committed:
Updates per CR, revert breaking changes (user output)

ankona committed:
conflict cleanup

ankona committed:
formatter

ankona committed:
fix non-conflict not caught in rebase

ankona committed:
Remove unnecessary pytest.ini

ankona committed:
fix inconsistent message capitalization

ankona committed:
Add fixture section to docstrings for modified tests

dafyddstephenson commented on cstar/roms/simulation.py:
```diff
@@ -1252,7 +1250,7 @@ def build(self, rebuild: bool = False) -> None:
             and not rebuild
         ):
-            print(
+            self.log.warning(
```
I think your previous choice of `info` was more appropriate here

ankona replied:
swapped

dafyddstephenson reviewed (changes requested):
I think there are literally two lines to change but I'd also just like to re-run the notebooks on this branch before approving. Should be good to go in sometime tomorrow morning

ankona committed:
Fix incorrect log level

ankona committed:
Merge branch 'main' into CW-782

Commit:
Update logging system: split stdout/stderr, change default format, fix logger persistence

- Modify default log format to be more concise and notebook-friendly.
- Split STDOUT/STDERR handling:
  - INFO and DEBUG logs routed to stdout via a custom handler.
  - WARNING and higher logs propagate to the root logger and stderr.
- Ensure root logger's handlers are explicitly set to WARNING level.
- Fix bug where _log attribute was pickled in Simulation.persist().
  - _log is now deleted during persist to ensure fresh logger on restore.
- Update and re-run example notebooks to reflect improved logging behavior.

dafyddstephenson commented:
See latest commit message for changes so far. Forgot to include in the message that `run_cmd` now also uses logging at the `DEBUG` level.

Hopefully the `ReviewNB` bot will show up here and allow for easier notebook review.

Still `TODO` on my end:
 - [ ] update the how-to guides section of the documentation
 - [x]  fix the unit tests

Commit:
fix unit tests and docstrings

Commit:
Merge branch 'main' into CW-782

Commit:
fix roms-tools python version in pyproject.toml

Commit:
begin updating how-to guides

dafyddstephenson committed:
update HPC how-to guide

dafyddstephenson commented:
This should now be good to go.
